### PR TITLE
STITCH-2584 Java SDK: Use bulk operations wherever possible on local database for sync

### PR DESCRIPTION
**Pull request by jsflax**

### CODE REVIEW

File: android/build.gradle [hunk 1]
```diff
@@ -32,7 +32,8 @@ subprojects {
                     if (file.exists()) {
                         properties.load(project.rootProject.file('local.properties').newDataInputStream())
                     }
-                    testInstrumentationRunnerArgument "test.stitch.baseURL", properties.getProperty("test.stitch.baseURL", "http://10.0.2.2:9090")
+                    testInstrumentationRunnerArgument "test.stitch.baseURL",
+                            properties.getProperty("test.stitch.baseURL", "http://10.1.14.181:9090")
```
Comment (edaniels) [inline android/build.gradle]: Undo

File: android/examples/stress-tests/build.gradle [hunk 2]
```diff
@@ -25,4 +35,9 @@ dependencies {
     implementation "com.android.support:appcompat-v7:${support_library_version}"
     implementation "com.android.support:recyclerview-v7:${support_library_version}"
     implementation "com.android.support.constraint:constraint-layout:1.1.0"
+//    implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-core:1.1.1'
```
Comment (edaniels) [inline android/examples/stress-tests/build.gradle]: uncomment?

File: android/services/mongodb-remote/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -1,2 +1,4 @@
-<manifest package="com.mongodb.stitch.android.services.mongodb.remote">
+<manifest package="com.mongodb.stitch.android.services.mongodb.remote"
+          xmlns:android="http://schemas.android.com/apk/res/android">
+    <application android:usesCleartextTraffic="true" />
```
Comment (edaniels) [inline android/services/mongodb-remote/src/main/AndroidManifest.xml]: I think you can undo these changes

File: android/testutils/src/main/java/com/mongodb/stitch/android/testutils/BaseStitchAndroidIntTest.kt [hunk 4]
```diff
@@ -36,7 +36,7 @@ open class BaseStitchAndroidIntTest : BaseStitchIntTest() {
     }
 
     override fun getStitchBaseURL(): String {
-        return InstrumentationRegistry.getArguments().getString("test.stitch.baseURL", "http://10.0.2.2:9090")
+        return "http://10.0.2.2:9090"
```
Comment (edaniels) [inline android/testutils/src/main/java/com/mongodb/stitch/android/testutils/BaseStitchAndroidIntTest.kt]: Should this be reverted?

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceSynchronizationConfig.java [hunk 5]
```diff
@@ -227,6 +236,32 @@ Codec getDocumentCodec() {
     return documentCodec;
   }
 
+  boolean addSynchronizedDocuments(
+      final BsonValue... documentIds
+  ) {
+    Map<BsonValue, CoreDocumentSynchronizationConfig> configs = new HashMap<>();
+    for (final BsonValue documentId : documentIds) {
+      if (getSynchronizedDocument(documentId) == null) {
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceSynchronizationConfig.java]: [q] Should we add a `getSynchronizedDocuments` to speed this up?

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceSynchronizationConfig.java [hunk 6]
```diff
@@ -250,6 +285,32 @@ boolean addSynchronizedDocument(
     }
   }
 
+  @Nullable
+  DeleteManyModel<CoreDocumentSynchronizationConfig> removeSynchronizedDocuments(
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceSynchronizationConfig.java]: [nit] Maybe could call this `prepareRemoveSynchronizedDocuments` so as to indicate this won't remove them at this call site

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/InstanceSynchronizationConfig.java [hunk 7]
```diff
@@ -162,6 +164,14 @@ public CoreDocumentSynchronizationConfig addAndGetSynchronizedDocument(
     return nsConfig.getSynchronizedDocument(documentId);
   }
 
+  @Nullable
+  DeleteManyModel<CoreDocumentSynchronizationConfig> removeSynchronizedDocuments(
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/InstanceSynchronizationConfig.java]: Similar comment about prepare

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 8]
```diff
@@ -91,6 +97,75 @@
  * documents.
  */
 public class DataSynchronizer implements NetworkMonitor.StateListener {
+  class BatchOps {
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: [nit] May be better to move this into its own class in a different file instead of this being an inner non-static class. Would just need to provide more args for its methods though to compensate not having direct access. Might be cleaner though.

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 9]
```diff
@@ -91,6 +97,75 @@
...
+    }
+
+    BatchOps() {
+      this.bulkWriteModels = new ArrayList<>();
+      this.configs = new ArrayList<>();
+      this.ids = new HashSet<>();
+    }
+
+    void merge(@Nullable final BatchOps batchOps) {
+      if (batchOps != null) {
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: [nit] Can make this if null early return to reduce indentation

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java [hunk 10]
```diff
@@ -2523,6 +2613,7 @@ public BsonValue apply(@NonNull final BsonDocument bsonDocument) {
           undoCollection.deleteOne(getDocumentIdFilter(documentId));
           eventsToEmit.add(event);
         }
+        triggerListeningToNamespace(namespace);
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: [q] Do we need to trigger here as a part of this change?

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java [hunk 11]
```diff
@@ -228,6 +267,16 @@ void setSomePendingWrites(
     }
   }
 
+  void setPendingWritesCompleteNoDB(final BsonDocument atVersion) {
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java]: [q] Is the point of this to make sync not skip over these during batch processing?

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java [hunk 12]
```diff
@@ -127,12 +143,12 @@ public boolean isStale() {
   public void setStale(final boolean stale) {
     docLock.writeLock().lock();
     try {
-      docsColl.updateOne(
-          getDocFilter(namespace, documentId),
-          new BsonDocument("$set",
-              new BsonDocument(
-                  CoreDocumentSynchronizationConfig.ConfigCodec.Fields.IS_STALE,
-                  new BsonBoolean(stale))));
+//      docsColl.updateOne(
```
Comment (edaniels) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java]: Should this be uncommented or method removed?

File: android/examples/stress-tests/build.gradle (hunk 2 above)
Comment (jsflax) [reply android/examples/stress-tests/build.gradle]: Can just delete, we don't need core.

File: android/testutils/src/main/java/com/mongodb/stitch/android/testutils/BaseStitchAndroidIntTest.kt (hunk 4 above)
Comment (jsflax) [reply android/testutils/src/main/java/com/mongodb/stitch/android/testutils/BaseStitchAndroidIntTest.kt]: Yes

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceSynchronizationConfig.java (hunk 5 above)
Comment (jsflax) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceSynchronizationConfig.java]: I don't think we need to, it's just reading from a hashmap.

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java (hunk 12 above)
Comment (jsflax) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java]: Removed

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java (hunk 10 above)
Comment (jsflax) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java]: Yes, `desync` used to trigger every time it was called. It no longer does that, so we need to do it manually after desyncing has been done. There is a cleaner way to do this by tracking what operations have been done and if the streams need to be restarted.

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java (hunk 11 above)
Comment (jsflax) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java]: The point of this is to not replace the doc config everytime we set pending writes– that's very slow. So we change it in memory here and do a bulk write after the pass on the namespace is done.

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceSynchronizationConfig.java (hunk 5 above)
Comment (edaniels) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceSynchronizationConfig.java]: :+1:

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java (hunk 11 above)
Comment (edaniels) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java]: cool

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java [hunk 13]
```diff
@@ -127,12 +140,6 @@ public boolean isStale() {
   public void setStale(final boolean stale) {
     docLock.writeLock().lock();
     try {
-      docsColl.updateOne(
-          getDocFilter(namespace, documentId),
-          new BsonDocument("$set",
-              new BsonDocument(
-                  CoreDocumentSynchronizationConfig.ConfigCodec.Fields.IS_STALE,
-                  new BsonBoolean(stale))));
       isStale = stale;
```
Comment (dkaminsky) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java]: [q] approve of caching locally but looks like this never gets written to the docConfig as a result - are we ok with this?

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java [hunk 14]
```diff
@@ -228,15 +235,34 @@ void setSomePendingWrites(
...
 
+  /**
+   * Sets that there are some pending writes that occurred at a time for an associated
+   * locally emitted change event. This variant updates the last version set.
+   *
+   * @param atTime      the time at which the write occurred.
+   * @param atVersion   the version for which the write occurred.
+   * @param changeEvent the description of the write/change.
+   */
+  void setSomePendingWrites(
```
Comment (dkaminsky) [inline core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java]: [q] why add this? does not appear this is called anywhere? did I miss it?

File: core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java (hunk 13 above)
Comment (jsflax) [reply core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java]: It does get written when we do the bulk writes.

In the event that the app crashes before it can be written, the document is marked stale anyway on app restart.
